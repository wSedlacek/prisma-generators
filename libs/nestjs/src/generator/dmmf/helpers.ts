// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
