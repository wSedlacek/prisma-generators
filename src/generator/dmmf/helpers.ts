// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
