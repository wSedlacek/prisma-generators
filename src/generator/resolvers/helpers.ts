import { getFieldTSType, getGraphQLType } from '../helpers';
import { DmmfDocument } from '../dmmf/dmmf-document';
import { DMMF } from '../dmmf/types';

export const generateCrudResolverClassMethodDeclaration = (
  action: DMMF.Action,
  typeName: string,
  method: DMMF.SchemaField,
  argsTypeName: string | undefined,
  collectionName: string,
  dmmfDocument: DmmfDocument,
  mapping: DMMF.Mapping
) => {
  const returnTSType = getFieldTSType(
    method.outputType,
    dmmfDocument,
    false,
    mapping.model,
    typeName
  );

  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${returnTSType}>`,
    decorators: [
      {
        name: `${action.operation}`,
        arguments: [
          `() => ${getGraphQLType(
            method.outputType,
            dmmfDocument,
            mapping.model,
            typeName
          )}`,
          `{
            ${[
              `nullable: ${!method.outputType.isRequired}`,
              `description: undefined`,
            ].join(',\n')}
          }`,
        ],
      },
    ],
    parameters: [
      {
        name: 'ctx',
        // TODO: import custom `ContextType`
        type: 'any',
        decorators: [{ name: 'Context', arguments: [] }],
      },
      ...(action.kind === 'aggregate'
        ? [
            {
              name: 'info',
              type: 'GraphQLResolveInfo',
              decorators: [{ name: 'Info', arguments: [] }],
            },
          ]
        : []),
      ...(!argsTypeName
        ? []
        : [
            {
              name: 'args',
              type: argsTypeName,
              decorators: [{ name: 'Args', arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === 'aggregate'
        ? [
            `const transformFields = (fields: Record<string, any>): Record<string, any> => {
              return Object.fromEntries(
                Object.entries(fields)
                  .filter(([key, value]) => !key.startsWith("_"))
                  .map<[string, any]>(([key, value]) => {
                    if (Object.keys(value).length === 0) {
                      return [key, true];
                    }
                    return [key, transformFields(value)];
                  }),
              );
            }`,
            `return ctx.prisma.${collectionName}.${action.kind}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : [
            `return plainToClass(${returnTSType
              .replace(/\[|\]/g, '')
              .replace(
                '| undefined',
                ''
              )}, await ctx.prisma.${collectionName}.${action.kind}(${
              argsTypeName ? 'args' : ''
            }) as ${getGraphQLType(
              method.outputType,
              dmmfDocument,
              mapping.model,
              typeName
            )});`,
          ],
  };
};
