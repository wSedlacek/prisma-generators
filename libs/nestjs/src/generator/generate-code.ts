import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { CompilerOptions, ModuleKind, Project, ScriptTarget } from 'ts-morph';

import { isDefined } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import { generateActionResolverClass } from './resolvers/separate-action';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

const baseCompilerOptions: CompilerOptions = {
  target: ScriptTarget.ES2019,
  module: ModuleKind.CommonJS,
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
};

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const emitTranspiledCode =
    options.emitTranspiledCode ??
    options.outputDirPath.includes('node_modules');
  const project = new Project({
    compilerOptions: {
      ...baseCompilerOptions,
      ...(emitTranspiledCode ? { declaration: true } : {}),
    },
  });
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  dmmfDocument.datamodel.enums.forEach((enumDef) =>
    generateEnumFromDef(project, baseDirPath, enumDef)
  );
  dmmfDocument.schema.enums
    // skip enums from datamodel
    .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
    .forEach((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef));
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);

  log('Generating models...');
  dmmfDocument.datamodel.models.forEach((model) =>
    generateObjectTypeClassFromModel(project, baseDirPath, model, dmmfDocument)
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  outputTypesToGenerate.forEach((type) =>
    generateOutputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    outputTypesFieldsArgsToGenerate.forEach((field) => {
      generateArgsTypeClassFromArgs(
        project,
        path.resolve(resolversDirPath, outputsFolderName),
        field.args,
        field.argsTypeName,
        dmmfDocument,
        2
      );
    });

    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
  }

  log('Generating input types...');

  dmmfDocument.schema.inputTypes.forEach((type) =>
    generateInputTypeClassFromType(
      project,
      resolversDirPath,
      type,
      dmmfDocument
    )
  );

  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );

  if (dmmfDocument.relationModels.length > 0) {
    log('Generating relation resolvers...');

    dmmfDocument.relationModels.forEach((relationModel) =>
      generateRelationsResolverClassesFromModel(
        project,
        baseDirPath,
        dmmfDocument,
        relationModel
      )
    );

    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      dmmfDocument.relationModels.map((relationModel) => ({
        resolverName: relationModel.resolverName,
        modelName: relationModel.model.typeName,
        hasSomeArgs: relationModel.relationFields.some(
          (field) => field.argsTypeName !== undefined
        ),
      }))
    );

    log('Generating relation resolver args...');
    dmmfDocument.relationModels.forEach((relationModelData) => {
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        relationModelData.model.typeName
      );

      relationModelData.relationFields
        .filter((field) => field.argsTypeName)
        .forEach((field) => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            field.outputTypeField.args,
            field.argsTypeName as string,
            dmmfDocument
          );
        });

      const argTypeNames = relationModelData.relationFields
        .filter((it) => it.argsTypeName !== undefined)
        .map((it) => it.argsTypeName)
        .filter(isDefined);

      if (argTypeNames.length) {
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
          undefined,
          { overwrite: true }
        );
        generateArgsBarrelFile(barrelExportSourceFile, argTypeNames);
      }
    });
  }

  log('Generating crud resolvers...');
  dmmfDocument.mappings.forEach((mapping) => {
    const model = dmmfDocument.datamodel.models.find(
      (it) => it.name === mapping.model
    ) as DMMF.Model;
    generateCrudResolverClassFromMapping(
      project,
      baseDirPath,
      mapping,
      model,
      dmmfDocument
    );
    mapping.actions.forEach((action) => {
      const actionModel = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      generateActionResolverClass(
        project,
        baseDirPath,
        actionModel,
        action,
        mapping,
        dmmfDocument
      );
    });
  });
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    dmmfDocument.mappings.map((mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;

      return {
        modelName: model.typeName,
        resolverName: mapping.resolverName,
        actionResolverNames: mapping.actions.map((it) => it.actionResolverName),
        hasSomeArgs: mapping.actions.some(
          (action) => action.argsTypeName !== undefined
        ),
      };
    })
  );

  log('Generating crud resolvers args...');
  dmmfDocument.mappings.forEach((mapping) => {
    const actionsWithArgs = mapping.actions.filter(
      (it) => it.argsTypeName !== undefined
    );

    if (actionsWithArgs.length) {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      ) as DMMF.Model;
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        model.typeName
      );
      actionsWithArgs.forEach((action) => {
        generateArgsTypeClassFromArgs(
          project,
          resolverDirPath,
          action.method.args,
          action.argsTypeName as string,
          dmmfDocument
        );
      });
      const barrelExportSourceFile = project.createSourceFile(
        path.resolve(resolverDirPath, argsFolderName, 'index.ts'),
        undefined,
        { overwrite: true }
      );
      generateArgsBarrelFile(
        barrelExportSourceFile,
        actionsWithArgs.map((it) => it.argsTypeName).filter(isDefined)
      );
    }
  });

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, dmmfDocument.relationModels.length > 0);

  log('Emitting generated code files');
  if (emitTranspiledCode) {
    await project.emit();
  } else {
    for (const file of project.getSourceFiles()) {
      file.formatText({ indentSize: 2 });
    }
    await project.save();
  }
};
