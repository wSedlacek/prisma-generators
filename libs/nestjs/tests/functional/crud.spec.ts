import { NestApplication } from '@nestjs/core';
import { GraphQLModule } from '@nestjs/graphql';
import { Test } from '@nestjs/testing';
import {
  ApolloServerTestClient,
  createTestClient,
} from 'apollo-server-testing';
import { promises as fs } from 'fs';
import 'reflect-metadata';

import { generateArtifactsDirPath } from '../helpers/artifacts-dir';
import { generateCodeFromSchema } from '../helpers/generate-code';
import { gql } from '../helpers/graphql-template';
import { prisma } from '../helpers/prisma-template';

describe('crud resolvers execution', () => {
  describe('basic operations', () => {
    let outputDirPath: string;

    let app: NestApplication;
    let apolloClient: ApolloServerTestClient;

    const findOneMock = jest.fn();
    const findManyMock = jest.fn();
    const createMock = jest.fn();
    const updateMock = jest.fn();
    const updateManyMock = jest.fn();
    const deleteMock = jest.fn();
    const deleteManyMock = jest.fn();
    const upsertMock = jest.fn();

    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath('functional-crud');
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = prisma`
        model User {
          intIdField          Int     @id @default(autoincrement())
          uniqueStringField   String  @unique
          optionalStringField String?
          dateField           DateTime
        }
      `;
      await generateCodeFromSchema(prismaSchema, { outputDirPath });
      const {
        UserCrudResolver,
      } = require(`${outputDirPath}/resolvers/crud/User/UserCrudResolver.ts`);

      const moduleFixture = await Test.createTestingModule({
        imports: [
          GraphQLModule.forRoot({
            autoSchemaFile: `${outputDirPath}/schema.graphql`,
            context: {
              prisma: {
                user: {
                  findOne: findOneMock,
                  findMany: findManyMock,
                  create: createMock,
                  delete: deleteMock,
                  deleteMany: deleteManyMock,
                  update: updateMock,
                  updateMany: updateManyMock,
                  upsert: upsertMock,
                },
              },
            },
          }),
        ],
        providers: [UserCrudResolver],
      }).compile();

      app = moduleFixture.createNestApplication();
      await app.init();

      const graphqlModule = moduleFixture.get(GraphQLModule);
      apolloClient = createTestClient((graphqlModule as any).apolloServer);
    }, 10000);

    it('should properly call PrismaClient on `findOne` action', async () => {
      findOneMock.mockResolvedValue({
        intIdField: 1,
        dateField: new Date('2019-12-31T14:16:02.572Z'),
      });

      const { data, errors } = await apolloClient.query({
        query: gql`
          query {
            user(where: { uniqueStringField: "uniqueValue" }) {
              intIdField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('user mocked response');
      expect(findOneMock.mock.calls).toMatchSnapshot('findOneUser call args');
    });

    it('should properly call PrismaClient on `findMany` action', async () => {
      findManyMock.mockResolvedValue([
        {
          intIdField: 1,
          dateField: new Date('2019-12-31T14:16:02.572Z'),
        },
      ]);

      const { data, errors } = await apolloClient.query({
        query: gql`
          query {
            users(
              take: 1
              skip: 1
              orderBy: { intIdField: desc }
              where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
            ) {
              intIdField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('users mocked response');
      expect(findManyMock.mock.calls).toMatchSnapshot('findManyUser call args');
    });

    it('should properly call PrismaClient on `create` action', async () => {
      createMock.mockResolvedValue({
        intIdField: 1,
        dateField: new Date('2019-12-31T14:16:02.572Z'),
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            createUser(
              data: {
                uniqueStringField: "unique"
                optionalStringField: "optional"
                dateField: "2019-12-31T14:16:02.572Z"
              }
            ) {
              intIdField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('createUser mocked response');
      expect(createMock.mock.calls).toMatchSnapshot('createUser call args');
    });

    it('should properly call PrismaClient on `update` action', async () => {
      updateMock.mockResolvedValue({
        intIdField: 1,
        dateField: '2019-12-31T14:16:02.572Z',
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            updateUser(
              data: { dateField: { set: "2019-12-31T14:16:02.572Z" } }
              where: { uniqueStringField: "unique" }
            ) {
              intIdField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('updateUser mocked response');
      expect(updateMock.mock.calls).toMatchSnapshot('updateUser call args');
    });

    it('should properly call PrismaClient on `updateMany` action', async () => {
      updateManyMock.mockResolvedValue({
        count: 3,
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            updateManyUser(
              data: { optionalStringField: null }
              where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
            ) {
              count
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('updateManyUser mocked response');
      expect(updateManyMock.mock.calls).toMatchSnapshot(
        'updateManyUser call args'
      );
    });

    it('should properly call PrismaClient on `delete` action', async () => {
      deleteMock.mockResolvedValue({
        intIdField: 1,
        dateField: new Date('2019-12-31T14:16:02.572Z'),
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            deleteUser(where: { uniqueStringField: "unique" }) {
              intIdField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('deleteUser mocked response');
      expect(deleteMock.mock.calls).toMatchSnapshot('deleteUser call args');
    });

    it('should properly call PrismaClient on `deleteMany` action', async () => {
      deleteManyMock.mockResolvedValue({
        count: 3,
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            deleteManyUser(
              where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
            ) {
              count
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('deleteManyUser mocked response');
      expect(deleteManyMock.mock.calls).toMatchSnapshot(
        'deleteManyUser call args'
      );
    });

    it('should properly call PrismaClient on `upsert` action', async () => {
      upsertMock.mockResolvedValue({
        intIdField: 1,
        uniqueStringField: 'unique',
        optionalStringField: 'optional',
        dateField: new Date('2019-12-31T14:16:02.572Z'),
      });

      const { data, errors } = await apolloClient.mutate({
        mutation: gql`
          mutation {
            upsertUser(
              where: { uniqueStringField: "unique" }
              create: {
                uniqueStringField: "unique"
                optionalStringField: "optional"
                dateField: "2019-12-31T14:16:02.572Z"
              }
              update: { optionalStringField: null }
            ) {
              intIdField
              uniqueStringField
              optionalStringField
              dateField
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('upsertUser mocked response');
      expect(upsertMock.mock.calls).toMatchSnapshot('upsertUser call args');
    });
  });

  describe('aggregations', () => {
    let outputDirPath: string;
    let app: NestApplication;
    let apolloClient: ApolloServerTestClient;

    const aggregateMock = jest.fn();

    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath('functional-crud');
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = prisma`
        model User {
          idField     Int  @id @default(autoincrement())
          intField    Int
          floatField  Int
        }
      `;
      await generateCodeFromSchema(prismaSchema, { outputDirPath });
      const {
        UserCrudResolver,
      } = require(`${outputDirPath}/resolvers/crud/User/UserCrudResolver.ts`);

      const moduleFixture = await Test.createTestingModule({
        imports: [
          GraphQLModule.forRoot({
            autoSchemaFile: `${outputDirPath}/schema.graphql`,
            context: {
              prisma: {
                user: {
                  aggregate: aggregateMock,
                },
              },
            },
          }),
        ],
        providers: [UserCrudResolver],
      }).compile();

      app = moduleFixture.createNestApplication();
      await app.init();

      const graphqlModule = moduleFixture.get(GraphQLModule);
      apolloClient = createTestClient((graphqlModule as any).apolloServer);
    });

    it('should properly call PrismaClient on `aggregate` action with simple count field', async () => {
      aggregateMock.mockResolvedValue({ count: 5 });

      const { data, errors } = await apolloClient.query({
        query: gql`
          query {
            aggregateUser(
              take: 1
              skip: 1
              orderBy: { intField: desc }
              where: { floatField: { lte: 50 } }
            ) {
              count
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('aggregateUserCount mocked response');
      expect(aggregateMock.mock.calls).toMatchSnapshot(
        'user.aggregate call args'
      );
    });

    it('should properly call PrismaClient on `aggregate` action with advanced operations', async () => {
      aggregateMock.mockResolvedValue({
        count: 2,
        min: {
          intField: 0,
          floatField: 0,
        },
        max: {
          intField: 10,
          floatField: 10,
        },
        sum: {
          intField: 10,
          floatField: 10,
        },
        avg: {
          intField: 5,
          floatField: 5,
        },
      });

      const { data, errors } = await apolloClient.query({
        query: gql`
          query {
            aggregateUser(
              take: 1
              skip: 1
              orderBy: { intField: desc }
              where: { floatField: { lte: 50 } }
            ) {
              __typename
              count
              min {
                __typename
                intField
                floatField
              }
              max {
                __typename
                intField
                floatField
              }
              sum {
                __typename
                intField
                floatField
              }
              avg {
                __typename
                intField
                floatField
              }
            }
          }
        `,
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot('aggregateUser mocked response');
      expect(aggregateMock.mock.calls).toMatchSnapshot(
        'user.aggregate call args'
      );
    });
  });
});
