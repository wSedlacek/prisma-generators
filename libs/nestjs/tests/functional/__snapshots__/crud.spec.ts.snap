// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with advanced operations: aggregateUser mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "__typename": "AggregateUser",
    "avg": Object {
      "__typename": "UserAvgAggregateOutputType",
      "floatField": 5,
      "intField": 5,
    },
    "count": 2,
    "max": Object {
      "__typename": "UserMaxAggregateOutputType",
      "floatField": 10,
      "intField": 10,
    },
    "min": Object {
      "__typename": "UserMinAggregateOutputType",
      "floatField": 0,
      "intField": 0,
    },
    "sum": Object {
      "__typename": "UserSumAggregateOutputType",
      "floatField": 10,
      "intField": 10,
    },
  },
}
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with advanced operations: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "count": true,
      "orderBy": Array [
        Object {
          "intField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": Object {
        "floatField": Object {
          "lte": 50,
        },
      },
    },
  ],
  Array [
    Object {
      "avg": Object {
        "floatField": true,
        "intField": true,
      },
      "count": true,
      "max": Object {
        "floatField": true,
        "intField": true,
      },
      "min": Object {
        "floatField": true,
        "intField": true,
      },
      "orderBy": Array [
        Object {
          "intField": "desc",
        },
      ],
      "skip": 1,
      "sum": Object {
        "floatField": true,
        "intField": true,
      },
      "take": 1,
      "where": Object {
        "floatField": Object {
          "lte": 50,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with simple count field: aggregateUserCount mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "count": 5,
  },
}
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with simple count field: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "count": true,
      "orderBy": Array [
        Object {
          "intField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": Object {
        "floatField": Object {
          "lte": 50,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`create\` action: createUser call args 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`create\` action: createUser mocked response 1`] = `
Object {
  "createUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`delete\` action: deleteUser call args 1`] = `
Array [
  Array [
    Object {
      "where": Object {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`delete\` action: deleteUser mocked response 1`] = `
Object {
  "deleteUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`deleteMany\` action: deleteManyUser call args 1`] = `
Array [
  Array [
    Object {
      "where": Object {
        "dateField": Object {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`deleteMany\` action: deleteManyUser mocked response 1`] = `
Object {
  "deleteManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findMany\` action: findManyUser call args 1`] = `
Array [
  Array [
    Object {
      "orderBy": Array [
        Object {
          "intIdField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": Object {
        "dateField": Object {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findMany\` action: users mocked response 1`] = `
Object {
  "users": Array [
    Object {
      "dateField": "2019-12-31T14:16:02.572Z",
      "intIdField": 1,
    },
  ],
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findOne\` action: findOneUser call args 1`] = `
Array [
  Array [
    Object {
      "where": Object {
        "uniqueStringField": "uniqueValue",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findOne\` action: user mocked response 1`] = `
Object {
  "user": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`update\` action: updateUser call args 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "dateField": Object {
          "set": 2019-12-31T14:16:02.572Z,
        },
      },
      "where": Object {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`update\` action: updateUser mocked response 1`] = `
Object {
  "updateUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`updateMany\` action: updateManyUser call args 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "optionalStringField": null,
      },
      "where": Object {
        "dateField": Object {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`updateMany\` action: updateManyUser mocked response 1`] = `
Object {
  "updateManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`upsert\` action: upsertUser call args 1`] = `
Array [
  Array [
    Object {
      "create": Object {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
      "update": Object {
        "optionalStringField": null,
      },
      "where": Object {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`upsert\` action: upsertUser mocked response 1`] = `
Object {
  "upsertUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
    "optionalStringField": "optional",
    "uniqueStringField": "unique",
  },
}
`;
